### PR TITLE
Add setQueryText method to be able to preset text to queryEditor

### PR DESCRIPTION
**Pull request by nshikov**
Need this to implement https://github.com/atom/fuzzy-finder/issues/279

### CODE REVIEW

**Comment by nshikov** [comment]
@as-cii hey, can you help?

File: src/select-list-view.js
Comment (as-cii) [inline src/select-list-view.js]: What do you think about renaming this method to `setQuery` and the parameter to `selectAll`?

Also, I think we should make the second parameter an object, so that it's clear what the boolean stands for, e.g.

```js
setQuery (text, {selectAll} = {})
```

**Review by as-cii — CHANGES_REQUESTED** [review]
Thanks for this pull-request, @nikshik! I think it'll make a pretty good addition to the API.

Other than the comment I left on the code, I think this needs a little more polish. In particular:

1. 🔥 Removing semicolons, to comply with the rest of the code standards.
2. 📝 Documenting this new API in README.md.

Thanks again for your contribution.

**Comment by nshikov** [comment]
Thanks for your response @as-cii, 
Sure will do, just not very clear on how do I document this in README.md as I only see that "props" Object passed to constructor is documented there

File: src/select-list-view.js
Comment (nshikov) [reply src/select-list-view.js]: I agree about setQuery. What do you think of:
`setQuery (text, selectText = false)`
Can do as object just not very clear how will it make more "clear what the boolean stands for"?

Comment (nshikov) [reply src/select-list-view.js]: ah ok, I got it you mean when the method called. I would stick to just boolean param, developer anyway have to know the API which he use. I also see similar signature already used in atom-select-list:
`selectIndex (index, updateComponent = true)`

Comment (as-cii) [reply src/select-list-view.js]: That's fair, although I would argue that the second parameter in `selectIndex` is an implementation detail and not really part of the API, so I would still prefer using an object if possible.

I like naming the parameter `selectText` as opposed to `selectAll`. 👍

**Comment by as-cii** [comment]
> Sure will do, just not very clear on how do I document this in README.md as I only see that "props" Object passed to constructor is documented there

Good point. I'll take care of writing the documentation after this pull-request gets merged. 👍

File: src/select-list-view.js
Comment (nshikov) [reply src/select-list-view.js]: @as-cii sorry, it's still not clear for me why you want it to be as object. For both variant developer will have to know the signature of the method:
`setQuery("test",true)`
`setQuery("test",{selectText:true})`
But in second case he will have to know the innner representation of the param and expose it outside the implementation and additionally calling setQuery method is much longer in second case.
Could you give me some details on this?

Comment (as-cii) [reply src/select-list-view.js]: I believe it helps clarify the purpose of the boolean to people that haven't written the code (or, after some months, to the person who wrote the code). Also, in case we add new parameters, they can just be added to the object as opposed to being passed as arguments (which would make the API intent even more obscure), e.g.

```js
selectList.setQueryText('hello', true, 1, 3)
// vs.
selectList.setQueryText(
  'hello', 
  {selectText: true, selectionStart: 1, selectionEnd: 3}
)
```

**Comment by nshikov** [comment]
@as-cii hey, could you please check it

**Review by as-cii — APPROVED** [review]

**Review by as-cii — CHANGES_REQUESTED** [review]
Looks great, nice work! ✨ 

One more thing: I have just noticed that there is no test for this new API, could you write one that exercises it so that we can have confidence it works as expected? Some scenarios worth testing in the same spec:

1. Calling `setQuery` passing `selectText: true`; this should set the text and select it too.
2. Calling `setQuery` without passing `selectText`; this should clear the previous selection and set the text.

Thanks!

**Comment by nshikov** [comment]
@as-cii Great you noticed missing test! Added test + one fix

**Comment by lexicalunit** [comment]
@as-cii 🤚 I would also like to have this functionality. This function could also be used to resolve #12 with the small addition an optional parameter to control if the call to `setQuery()` should trigger a query change callback or not.

```diff
  constructor (props) {
  [...]
-    this.disposables.add(this.refs.queryEditor.onDidChange(this.didChangeQuery.bind(this)))
+    this.didChangeQueryDisposable = this.refs.queryEditor.onDidChange(this.didChangeQuery.bind(this))
  [...]
  }
```

```diff
-  setQuery (text, {selectText} = {}) {
+  setQuery (text, {selectText, doDidChangeQuery} = {}) {
+    if (doDidChangeQuery === false) {
+      this.didChangeQueryDisposable.dispose()
+    }
     let queryEditor = this.refs.queryEditor
     queryEditor.setText(text)
     if (selectText) {
       queryEditor.selectAll()
     } else {
       let lastSel = queryEditor.getLastSelection();
       if (lastSel) {
         lastSel.clear();
       }
     }
+    if (doDidChangeQuery === false) {
+      this.didChangeQueryDisposable = queryEditor.onDidChange(this.didChangeQuery.bind(this))
+    }
   }
```

```diff
   destroy () {
     this.disposables.dispose()
+    this.didChangeQueryDisposable.dispose()
     return etch.destroy(this)
   }
```

@nikshik You probably also want to remove the `;` from lines `let lastSel = queryEditor.getLastSelection();` and `lastSel.clear();`. That would better match the style in the rest of this codebase.

**Comment by nshikov** [comment]
Hey @as-cii, should I do anything else for this pr?

**Comment by as-cii** [comment]
Hey @nikshik, sorry for the delay. This looks good! I'd like to make some further tweaks to this branch, so can you grant me push access to it? Here's a guide on how to do it: https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/.

Thanks!

**Review by as-cii — APPROVED** [review]

**Comment by as-cii** [comment]
Thanks again for this, really nice work @nikshik! 

I ended up moving the `setQuery` logic inside of `update`, so that it can be used from other etch views as well. This has been published as v0.5.0.

**Comment by nshikov** [comment]
Great! Thanks for your help and approval @as-cii!

**Pull request merged**
